validation: don't re-acquire cs_main during IBD in CChainState::IsInitialBlockDownload()

jonatack opened the pull request:
No longer re-acquire lock `cs_main` during IBD in `CChainState::IsInitialBlockDownload()`, as most of its callers already hold the lock. Have these callers invoke non-locking function `CChainState::IsIBD()` instead.  Apply thread safety analysis to the IBD functions. Retains the post-IBD optimization latch for the callers that do not hold the lock. Makes explicit which callers hold `cs_main`, and which do not.

jonatack commented:
Noticed while reviewing #24178.

sdaftuar commented:
I recall that the reason IsIBD was written like this:

```
bool CChainState::IsInitialBlockDownload() const
{
    // Optimization: pre-test latch before taking the lock.
    if (m_cached_finished_ibd.load(std::memory_order_relaxed))
        return false;

    LOCK(cs_main);
```

is so that in the common case -- which is that we've already latched to false -- that we can respond without ever acquiring cs_main inside this function. 

Maybe there's a better refactor that could be done so that we can reduce locking contention on cs_main further, but this patch would have all the call sites grab the lock in order to even call IsIBD(), which would add more cs_main lock acquisitions in general -- and that seems counter to where we want to end up?

Is the motivation for this patch to reduce the uses of cs_main as a recursive mutex, and thereby get to a better design?  If so that sounds like a good direction, but I think it'd be better if we avoided adding extra lock acquisitions (and holding cs_main longer) in order to get there.  For instance it seems like with this patch some of the validation callbacks will now be invoked while holding cs_main the whole time?  That seems undesirable.

jonatack commented:
Yes, the trade-off this pull makes is foregoing the optimization for a half dozen callers that don't already hold cs_main, in favor of avoiding re-taking the cs_main lock for the 18 callers that do already hold it, by moving the lock-taking to the half dozen callers.

sdaftuar commented on src/validation.cpp:
```diff
@@ -3111,13 +3111,13 @@ bool CChainState::InvalidateBlock(BlockValidationState& state, CBlockIndex* pind
-        uiInterface.NotifyBlockTip(GetSynchronizationState(IsInitialBlockDownload()), to_mark_failed->pprev);
+        // Only notify about a new block tip if the active chain was modified.
+        if (pindex_was_in_chain) {
+            uiInterface.NotifyBlockTip(GetSynchronizationState(IsInitialBlockDownload()), to_mark_failed->pprev);
```
I believe this change will cause the NotifyBlockTip() callbacks to be invoked while holding cs_main the whole time.

jonatack replied:
Thanks! Narrowed the lock to the IBD call: `uiInterface.NotifyBlockTip(GetSynchronizationState(WITH_LOCK(::cs_main, return IsInitialBlockDownload())), to_mark_failed->pprev);`

jamesob commented:
> For instance it seems like with this patch some of the validation callbacks will now be invoked while holding cs_main the whole time? That seems undesirable.

Agree with @sdaftuar; IsIBD() is called in a few performance-critical places and at the very least this change should be benchmarked before merge. I'm Concept NACK if this change is essentially just for refactoring's sake; if there's some overarching benefit it would be nice to hear that.

jonatack commented:
@jamesob I agree; potentially improving performance by not needlessly retaking the cs_main lock is the only point here.

sdaftuar commented:
> potentially improving performance by not needlessly retaking the cs_main lock is the only point here

AFAIU lock acquisition is essentially free if it's already held, so I would be surprised if there's a performance gain from avoiding re-acquisition.

jonatack commented:
Looking ahead at the assumeutxo work in the pipeline, IsIBD() will see further callers already holding cs_main. If the assumption holds that a thread safety annotation is preferable to retaking a lock, we might have our cake and eat it too: the majority can call the core IsIBD() with an annotation, and the seven callers sans lock can go through a wrapper with optimization and lock.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -4608,11 +4609,11 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
     {
         LOCK(cs_main);
 
+        MaybeSendAddr(*pto, *peer, current_time);
```
Slightly tend toward NACK. We shouldn't be broadening the scope of cs_main, but reduce it. Seems wrong to use a validation lock for sending p2p addresses on the network.

jonatack replied:
Agree. Addressed in latest push.

MarcoFalke commented:
This also has a silent merge conflict with #22053

jonatack committed:
validation: don't re-acquire cs_main in IsInitialBlockDownload() during IBD

as most of its callers already hold the cs_main lock;
have these callers invoke non-locking IsIBD() instead.

Retains the post-IBD optimization latch for the callers that
do not hold the lock.

Makes explicit which callers hold cs_main and which do not.

jonatack commented:
Pushed a different approach along the lines of https://github.com/bitcoin/bitcoin/pull/24220#issuecomment-1026298608. It does not increase the scope of `cs_main` and maintains the optimization, while avoiding reacquiring the mutex for the majority of callers that already hold it.

jonatack commented:
Simplified to one commit to facilitate review. Will propose the other changes separately.

MarcoFalke commented:
> to reduce the use of cs_main as a recursive mutex.

I think this is a goal that is almost impossible to achieve with a mutex visible in the global scope/namespace. And if someone did it, it would be brittle and easily lead to UB.

See also how for a far simpler and private mutex (`m_chainstate_mutex`), it took two attempts to get right: #24235.

Personally, I think a better way would be to no longer hold cs_main for non-validation uses and then make cs_main private to chainman.

jonatack commented:
Yes. This change clarifies which callers hold cs_main and which do not, now and going forward.

luke-jr commented on src/validation.h:
```diff
@@ -676,7 +676,13 @@ class CChainState
 
     /** Check whether we are doing an initial block download (synchronizing from disk or network) */
-    bool IsInitialBlockDownload() const;
+    bool IsInitialBlockDownload() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
Would prefer to make `IsIBD` the new semantics

jonatack replied:
Done

PastaPastaPasta commented on src/validation.h:
```diff
@@ -676,7 +676,13 @@ class CChainState
+     * Same as `IsInitialBlockDownload()`, but also locks mutex cs_main during
+     * initial block download. Use for callers not already holding the lock.
+     */
+    bool IsIBD() const;
```
should we add a `LOCKS_EXCLUDED` to prevent accidentally calling this when we know we have cs_main locked?

jonatack replied:
You read my mind. Added an `LOCK_EXCLUDED` thread safety annotation to the declaration along with an `AssertLockNotHeld(::cs_main);` run-time assertion in the definition.

jonatack committed:
validation: apply thread safety analysis to IBD boolean functions

Ensures that callers now and going forward invoke the correct function.

jonatack committed:
validation: create non-locking IsIBD() wrapped by IsInitialBlockDownload()

jonatack commented:
Updated with @luke-jr and @PastaPastaPasta feedback (thanks!)

MarcoFalke commented:
Closing for now due to inactivity. Let us know if this should be reopened.
